File: RIddhi/productofarrayexceptself.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        product = 1
        zeros_count = 0
        for n in nums:
            if n == 0:
                zeros_count += 1
            else:
                product *= n
            
        array = []
        for i in range(len(nums)):
            if (zeros_count >= 2):
                array.append(0)
            elif (zeros_count == 1):
                if (nums[i] == 0):
                    array.append(product)
                else:
                    array.append(0)
            else:
                array.append(product//nums[i])

        return array

File: Christine/productofarrayexceptself.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int[] answer = new int[nums.length];
        answer[0] = 1;

        int[] prefix = new int[nums.length];
        prefix[0] = 1;

        for (int i = 1; i < nums.length; i++) {
            answer[i] = answer[i - 1] * nums[i - 1];
        }

        int multiplier = 1;
        for (int i = answer.length - 1; i >= 0; i--) {
            answer[i] = answer[i] * multiplier;
            multiplier = multiplier * nums[i];
        }

        return answer;
    }
}
